### ISSUE TITLE
Upgrade Spring Data modules to the Spring Data Hopper release

### ISSUE THREAD

**Issue by jxblum** [spring-projects/spring-session#470]
This effort will update _Spring Session's_ dependency on _Spring Data GemFire_ to `1.8.1.RELEASE`.

**Comment by jxblum** [comment]
Seems the **new**, upgraded _Spring Data GemFire_ version (i.e. `1.8.1.RELEASE`) is adversely affecting the _Spring Session Data Mongo_ tests.

Mongo tests are failing with the following, for example...

```
org.springframework.session.data.mongo.MongoRepositoryJdkSerializationITests > findByPrincipalNameNoPrincipalNameChange FAILED
    java.lang.IllegalStateException
        Caused by: org.springframework.beans.factory.UnsatisfiedDependencyException
            Caused by: org.springframework.beans.factory.BeanCreationException
                Caused by: org.springframework.beans.BeanInstantiationException
                    Caused by: java.lang.NoSuchFieldError
```

Upon closer inspection (via the surefire test reports), the actual exception associated with the test failure is...

```
Caused by: java.lang.NoSuchFieldError: INSTANCE
	at org.springframework.data.mongodb.core.convert.DefaultMongoTypeMapper.<init>(DefaultMongoTypeMapper.java:65)
	at org.springframework.data.mongodb.core.convert.MappingMongoConverter.<init>(MappingMongoConverter.java:111)
	at org.springframework.data.mongodb.core.MongoTemplate.getDefaultMongoConverter(MongoTemplate.java:2045)
	at org.springframework.data.mongodb.core.MongoTemplate.<init>(MongoTemplate.java:217)
	at org.springframework.data.mongodb.core.MongoTemplate.<init>(MongoTemplate.java:181)
	at org.springframework.session.data.mongo.AbstractMongoRepositoryITests$BaseConfig.mongoOperations(AbstractMongoRepositoryITests.java:392)
	at org.springframework.session.data.mongo.MongoRepositoryJdkSerializationITests$Config$$EnhancerBySpringCGLIB$$ae59f44e.CGLIB$mongoOperations$1(<generated>)
	at org.springframework.session.data.mongo.MongoRepositoryJdkSerializationITests$Config$$EnhancerBySpringCGLIB$$ae59f44e$$FastClassBySpringCGLIB$$c17bd910.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:228)
	at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:355)
	at org.springframework.session.data.mongo.MongoRepositoryJdkSerializationITests$Config$$EnhancerBySpringCGLIB$$ae59f44e.mongoOperations(<generated>)
	at sun.reflect.GeneratedMethodAccessor38.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:162)
	... 77 more
```

Because _Spring Data GemFire_ `1.8.1.RELEASE`, [by association](https://github.com/spring-projects/spring-data-gemfire/blob/1.8.1.RELEASE/pom.xml#L12-L16) to the SD parent, has the same [dependency on SD Commons](https://github.com/spring-projects/spring-data-gemfire/blob/1.8.1.RELEASE/pom.xml#L39-L43) as _Spring Data MongoDB_ ([here](https://github.com/spring-projects/spring-data-mongodb/blob/master/spring-data-mongodb/pom.xml#L55-L59)), by upgrading _Spring Data GemFire_ to `1.8.1.RELEASE` [in _Spring Session_](https://github.com/spring-projects/spring-session/blob/master/gradle.properties#L21), that is effectively pulling in a [version of SD Commons](https://github.com/spring-projects/spring-data-gemfire/blob/1.8.1.RELEASE/pom.xml#L24) (`1.12.1.RELEASE`) that is incompatible with the [version of SD Commons](https://github.com/spring-projects/spring-data-mongodb/blob/1.8.2.RELEASE/spring-data-mongodb/pom.xml#L58) (i.e. `1.11.2.RELEASE`; determined by [this](https://github.com/spring-projects/spring-data-mongodb/blob/1.8.2.RELEASE/spring-data-mongodb/pom.xml#L15), [here](https://github.com/spring-projects/spring-data-mongodb/blob/1.8.2.RELEASE/pom.xml#L31)) based on the [version of _Spring Data MongoDB_](https://github.com/spring-projects/spring-session/blob/master/gradle.properties#L19) (`1.8.2.RELEASE`) currently used by _Spring Session_.

Indeed, we can see that the **field** in question (i.e. "`INSTANCE`") from the _SD Commons_ class `SimpleTypeInformationMapper` existed in _SD Commons_ `1.11.2.RELEASE` ([here](https://github.com/spring-projects/spring-data-commons/blob/1.11.2.RELEASE/src/main/java/org/springframework/data/convert/SimpleTypeInformationMapper.java#L35)), the version used by _SD MongoDB_ `1.8.2.RELEASE` used currently by _Spring Session_, but is not in _SD Commons_ `1.12.1.RELEASE` ([here](https://github.com/spring-projects/spring-data-commons/blob/1.12.1.RELEASE/src/main/java/org/springframework/data/convert/SimpleTypeInformationMapper.java)), the version expected by the SDG `1.8.1.RELEASE` upgrade.

**Comment by jxblum** [comment]
There is 2 ways to fix this...

1. Add a <dependencyManagement> section to Spring Session to control the version of SD Commons, or...

2. Better yet, upgrade both SD _MongoDB_ and SD _Redis_ to Hopper versions in _Spring Session_.

The later is recommended.  The dependencies could probably better be controlled with either _Spring Boot_ or using the curated set of dependencies in the Spring IO Platform.  However, the later (unlike _Spring Boot_) progresses at a slower pace.

**Comment by rwinch** [comment]
@jxblum Feel free to update dependencies as necessary. I'd prefer to consistently use the versions in the Spring IO platform as we must be compliant with that anyways.

**Comment by jxblum** [comment]
@rwinch Thanks.  I was just looking into the _Spring IO Platform_ curated/harmonized set of dependencies as specified by the [BOM](https://github.com/spring-io/platform/blob/master/platform-bom/pom.xml) and noticed that the _Spring IO Platform_ [inherits from the _Spring Boot Starter Parent_](https://github.com/spring-io/platform/blob/master/platform-bom/pom.xml#L6-L11), with only minimal (Spring-based) [additions](https://github.com/spring-io/platform/blob/master/platform-bom/pom.xml#L48-L54), i.e. no "overrides".

Since the _Spring IO Platform_ is **inheriting** from the now, current _Spring Boot_ **1.4** runtime, most of the dependencies are going to be defined [here](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-dependencies/pom.xml), and _Spring Boot_ 1.4 is [based on _Spring Data_ **Hopper**](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-dependencies/pom.xml#L146) already.

Therefore, I think it is OK if we bump the versions of all SD modules supported by _Spring Session_.

Let me know if you have additional thoughts or concerns.

Thanks.

**Comment by rwinch** [comment]
@jxblum I think it is fine to bump the Spring Data versions to Hopper. Thanks!

**Comment by jxblum** [comment]
@rwinch One more thing...

I just observed that _Spring Session_ is still [based on](https://github.com/spring-projects/spring-session/blob/master/build.gradle#L21) _Spring Boot_ `1.3.2.RELEASE` (current GA is `1.3.3.RELEASE`).  `1.4 M1` is also available too, so it may just boil down to timing between _Spring Boot_ **1.4** GA and the next _Spring Session_ release.

The only reason I mention it is, there are differences between _Spring Boot_ 1.3 SD modules and 1.4 (namely Gosling vs. Hopper), the whole reason for the mess with updating SDG in Spring Session, causing conflicts for Mongo support.

Anyway, just an FYI.

**Comment by rwinch** [comment]
I'm not sure when Spring Boot GA is scheduled, but given there is no RC I think Spring Session will be done before then. We are in the RC stage.

If you can verify everything works with Hopper in master, then go ahead and update. We just need to ensure that our Spring IO checks don't break (i.e. ensure you have JDK7_HOME setup in gradle.properties and run springIoCheck)

**Comment by jxblum** [comment]
@rwinch 

> ensure you have JDK7_HOME setup in gradle.properties

I am not sure what you mean by having `JDK7_HOME` setup in `gradle.properties`.  I did just run the `springIoCheck` on _Spring Session_ with success...

```
$ gradlew springIoCheck
Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=512M; support was removed in 8.0
:spring-session:springIoAlternativeDependenciesCheck
:spring-session:springIoDependencyVersionMappingCheck
:spring-session:springIoIncompleteExcludesCheck
:spring-session:springIoTest UP-TO-DATE
:spring-session:springIoCheck
:spring-session-data-gemfire:springIoAlternativeDependenciesCheck
:spring-session-data-gemfire:springIoDependencyVersionMappingCheck
:spring-session-data-gemfire:springIoIncompleteExcludesCheck
:spring-session-data-gemfire:springIoTest UP-TO-DATE
:spring-session-data-gemfire:springIoCheck
:spring-session-data-mongo:springIoAlternativeDependenciesCheck
:spring-session-data-mongo:springIoDependencyVersionMappingCheck
:spring-session-data-mongo:springIoIncompleteExcludesCheck
:spring-session-data-mongo:springIoTest UP-TO-DATE
:spring-session-data-mongo:springIoCheck
:spring-session-data-redis:springIoAlternativeDependenciesCheck
:spring-session-data-redis:springIoDependencyVersionMappingCheck
:spring-session-data-redis:springIoIncompleteExcludesCheck
:spring-session-data-redis:springIoTest UP-TO-DATE
:spring-session-data-redis:springIoCheck
:spring-session-jdbc:springIoAlternativeDependenciesCheck
:spring-session-jdbc:springIoDependencyVersionMappingCheck
:spring-session-jdbc:springIoIncompleteExcludesCheck
:spring-session-jdbc:springIoTest UP-TO-DATE
:spring-session-jdbc:springIoCheck
```

**Comment by jxblum** [comment]
@rwinch I have successfully completed this PR.

All SD modules have been updated to the [Hopper](https://github.com/spring-projects/spring-data-commons/wiki/Release-Train-Hopper#participating-modules) release.  I had to update the **Jedis** driver and exclude the **SLF4J** dependencies in SD _Redis_ to prevent conflicts when running in Tomcat.

All Spring Session tests ran successfully (ran with `$gradlew clean build install`) and all Spring IO Platform checks passed.

If you have any questions or concerns, let me know.

Thanks.
